### PR TITLE
🏗 Add support for #alias imports

### PR DESCRIPTION
**Pull request by rcebulko**
See #34529 (I2I)
Useful explanatory comment: https://github.com/ampproject/amphtml/pull/34562#discussion_r640747449

Note: There is an eslint plugin [`eslint-plugin-module-resolver`](https://github.com/HeroProtagonist/eslint-plugin-module-resolver) that auto-corrects imports to use aliases where available. The repo appears to be unmaintained, with out-of-date packages and quite some time since the last update. There are a couple bugs/annoyances in the current repo which I have patched in a local fork:
1. The code as-written requires a `.babelrc` file to extract the alias map; I have patched this to allow providing the alias map directly as a plugin option in `.eslintrc.js`
2. The code as-written doesn't handle string manipulations properly (if `#service/*` => `src/service/*`, `../../src/services` currently is "fixed" to `#service/s`); I have patched this to handle the end-of-path check properly.

This PR introduces support for aliased paths and a partial update of `#core` imports to introduce the syntax and start verifying there are no issues with development. Meanwhile, I plan to pursue foring the `eslint-plugin-module-resolver` plugin repo under the `ampproject` org, updating its packages, patching the bugs, and publishing to NPM so we can eventually rely on it. In the interim, it can be used locally to create PRs updating the codebase to use the import paths.

Commit containing all the automatic lint fixes: https://github.com/ampproject/amphtml/commit/af446f7aec341bc0cac7f510aeafb71c9aa93b2b

### CODE REVIEW

**Comment by rcebulko** [comment]
"real" diff excluding updates to imports: 9bb603c58...ef0262a12

File: package.json [hunk 1]
```diff
@@ -90,13 +91,15 @@
...
     "eslint-config-prettier": "8.3.0",
+    "eslint-import-resolver-babel-module": "5.3.1",
     "eslint-plugin-chai-expect": "2.2.0",
     "eslint-plugin-eslint-plugin": "3.0.3",
     "eslint-plugin-google-camelcase": "0.0.2",
     "eslint-plugin-import": "2.23.3",
     "eslint-plugin-jasmine": "4.1.2",
     "eslint-plugin-jsdoc": "35.0.0",
     "eslint-plugin-local": "1.0.0",
+    "eslint-plugin-module-resolver": "git+https://github.com/rcebulko/eslint-plugin-module-resolver.git",
```
Comment (rcebulko) [inline package.json:102]: This one (which I have patched to fix a bug, waiting to see if the repo is active) is for eslint to require/auto-magically use alias imports when available. An example of the corresponding changes for `src/preact/component/3p-frame.js` is below. You can see that
- imports with available aliases are shortened
- module imports (`./iframe` for example) are left untouched
- un-aliased imports (`../../3p-frame`) are left untouched

```diff
-import * as Preact from '../../../src/preact';
+import * as Preact from '#preact';
 import {IframeEmbed} from './iframe';
-import {dict} from '../../core/types/object';
-import {forwardRef} from '../compat';
+import {dict} from '#core/types/object';
+import {forwardRef} from '#preact/compat';
 import {
   generateSentinel,
   getBootstrapUrl,
   getDefaultBootstrapBaseUrl,
 } from '../../3p-frame';
-import {
-  getOptionalSandboxFlags,
-  getRequiredSandboxFlags,
-} from '../../core/3p-frame';
-import {includes} from '../../core/types/string';
+import {getOptionalSandboxFlags, getRequiredSandboxFlags} from '#core/3p-frame';
+import {includes} from '#core/types/string';
 import {parseUrlDeprecated} from '../../url';
-import {sequentialIdGenerator} from '../../core/math/id-generator';
+import {sequentialIdGenerator} from '#core/math/id-generator';
 import {
   useEffect,
   useImperativeHandle,
@@ -37,7 +34,7 @@ import {
   useMemo,
   useRef,
   useState,
-} from '../../../src/preact';
+} from '#preact';
```

File: .babelrc.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This is consumed only by the linter, which relies on plugins
```
[the remark this answers is not in the document]
Comment (HeroProtagonist) [reply .babelrc.js]: It is maintained :) I am looking at your PR now

**Comment by zargold** [comment]
Yes that esmodule lint/resolver is good stuff.

File: build-system/babel-config/import-resolver.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const JSCONFIG_PATH = path.join(__dirname, '../../jsconfig.json');
```
Comment (jridgewell) [inline build-system/babel-config/import-resolver.js:21]:
```suggestion
const JSCONFIG_PATH = path.join(__dirname, '..', '..', 'jsconfig.json');
```

**Review by jridgewell — APPROVED** [review]

**Review by kristoferbaxter — APPROVED** [review]
Looks good overall, only concerns are around jsconfig. 

Should we (in the future) have a check that validates the config contains all of the hashtag prefix paths used in the project?

**Comment by rcebulko** [comment]
> Looks good overall, only concerns are around jsconfig.
> 
> Should we (in the future) have a check that validates the config contains all of the hashtag prefix paths used in the project?

I'd say that if such an path were used, it should fail other tests or the build (as any nonsense path should). Pushing past that, down the line I'm looking to enable import rules in eslint to detect things like invalid imports, unused exports, etc.

**Comment by kristoferbaxter** [comment]
> I'd say that if such an path were used, it should fail other tests or the build (as any nonsense path should). Pushing past that, down the line I'm looking to enable import rules in eslint to detect things like invalid imports, unused exports, etc.

This sounds ok to me, so long as the culprit is easy to identify and the CI time to execute and fail in this manner is short.

**Comment by rcebulko** [comment]
> This sounds ok to me, so long as the culprit is easy to identify and the CI time to execute and fail in this manner is short.

Eslint rules will definitely help with the faster turnaround
FWIW i also set jsconfig OWNERS to require a review from wg-performance, so that file should have fairly minimal churn

**Pull request merged**